lib/api/BackbeatServer.js

```diff
@@ -0,0 +1,233 @@
+     * @param {object} req - request object
+     * @param {object} res - response object
+     * @return {undefined}
+     */
+    _requestListener(req, res) {
+        req.socket.setNoDelay();
+        const bbRequest = new BackbeatRequest()
+            .setRequest(req)
+            .setResponse(res)
+            .setLog(this._logger.newRequestLogger())
+            .setRoute(req.url);
+
+        // check request conditions and all 3rd-party conditions here
+        // should ISRs always be checked as well?
+        // What to name `_areConditionsOk`?
+        if (this._isValidRequest(req, bbRequest)
+        && this._areConditionsOk(bbRequest)) {
+            bbRequest.setStatusCode(200);
+            bbRequest.setRoute(bbRequest.getRoute().slice(3));
+
+            this.backbeatAPI[bbRequest.getRoute()](bbRequest, this._response);
+        }
+    }
+
+    /**
+     * Server's response to the client
+     * @param {BackbeatRequest} backbeatRequest - Backbeat request object
+     * @param {object} data - response to send to client
+     * @return {object} res - response object
+     */
+    _response(backbeatRequest, data) {
+        const log = backbeatRequest.getLog();
+        const req = backbeatRequest.getRequest();
+        const res = backbeatRequest.getResponse();
+        log.trace('writing HTTP response', {
+            method: 'BackbeatServer._response',
+        });
+        const code = backbeatRequest.getStatusCode();
+        const payload = Buffer.from(JSON.stringify(data), 'utf8');
+
+        res.writeHead(code, {
+            'content-type': 'application/json',
+            'content-length': payload.length,
+        });
+        res.write(payload);
+        BackbeatServer.logRequestEnd(log, req, res);
+        return res.end();
+    }
+
+    /**
+     * start BackbeatServer
+     * @return {undefined}
+     */
+    start() {
+        this.server = http.createServer((req, res) => {
+            this._requestListener(req, res);
+        });
+
+        this.server.on('listening', () => {
+            const addr = this.server.address() || {
+                address: '0.0.0.0',
+                port: this._port,
+            };
+            this._logger.trace('server started', {
+                address: addr.address,
+                port: addr.port,
+                pid: process.pid,
+            });
+        });
+        this.server.listen(this._port);
+    }
+
+    /*
+     * stop BackbeatServer and exit running process properly
+     */
+    stop() {
+        this._logger.info(`worker ${this.worker} shutting down`);
+        this.server.close();
+        process.exit(0);
+    }
+}
+
+/**
+ * start - start the backbeat API server
+ * @param {object} config - configurations required
+ * @return {undefined}
+ */
+function run(config) {
+    const logger = new Logger('BackbeatServer');
+    const loggerAPI = new Logger('BackbeatAPI');
+    const workers = os.cpus().length;
+    const cluster = new Clustering(workers, logger);
```

Comment:
Using clustering for this server sounds like a premature optimization to me, I'm not sure we will ever have CPU intensive operations in this server that would require multiple CPU assignment. We may actually not want to allocate too much resources for an admin/monitoring interface so that other processes can benefit more, but if we plan to plug actual heavy work into this server later, why not. @rahulreddy?